NomicFoundation/hardhat-vscode
# feat: add support for region comments

- [ ] Because this PR includes a **bug fix**, relevant tests have been included.
- [x] Because this PR includes a **new feature**, the change was previously discussed on an Issue or with someone from the team.
- [ ] I didn't do anything of this.

---
## Overview

This PR introduces support for region comments within Solidity `.sol` files previously mentioned in https://github.com/NomicFoundation/hardhat-vscode/issues/452, addressing the issue where developers were seeking the ability to organize code into collapsible sections for improved readability and navigation. 

## Changes

- Modified the Solidity language configuration file to recognize the `//#region` and `//#endregion` patterns.

## Implementation Details

Region comments are implemented by modifying the Solidity language grammar to include start and end region markers. The pattern follows the typical convention of `//#region` and `//#endregion` which is familiar to developers from other programming languages.

```solidity
//#region ExampleRegion

// Your Solidity code here

//#endregion ExampleRegion
```

The language configuration now includes logic to parse these markers and translate them into foldable regions within VS Code.

## Screenshots

_Included are screenshots demonstrating the collapsible regions in action within a Solidity file._

<img width="809" alt="image" src="https://github.com/NomicFoundation/hardhat-vscode/assets/4002635/208ff6f1-fd64-4538-bb9f-7246f9e00fc8">

\
<img width="862" alt="image" src="https://github.com/NomicFoundation/hardhat-vscode/assets/4002635/e00dfcea-fc38-4fd0-88dc-803a9a779e9e">

## How to test this feature

1. Check out this PR and follow the instructions in [CONTRIBUTING.md#running-locally](https://github.com/NomicFoundation/hardhat-vscode/blob/development/CONTRIBUTING.md#running-locally) to run it locally.
2. Open a `.sol` file in VS Code.
3. Insert `//#region` and `//#endregion` comments around the code you wish to collapse.
4. Verify that you can collapse and expand the region as expected.
